presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java

```diff
@@ -0,0 +1,227 @@
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.optimizations.SymbolMapper;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.FINAL;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.PARTIAL;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+
+public class PushPartialAggregationThroughExchange
+        implements Rule
+{
+    private static final Pattern PATTERN = Pattern.typeOf(AggregationNode.class);
+
+    private final FunctionRegistry functionRegistry;
+
+    public PushPartialAggregationThroughExchange(FunctionRegistry functionRegistry)
+    {
+        this.functionRegistry = functionRegistry;
+    }
+
+    @Override
+    public Pattern getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Context context)
+    {
+        AggregationNode aggregationNode = (AggregationNode) node;
+
+        PlanNode childNode = context.getLookup().resolve(aggregationNode.getSource());
+        if (!(childNode instanceof ExchangeNode)) {
+            return Optional.empty();
+        }
+
+        ExchangeNode exchangeNode = (ExchangeNode) childNode;
+
+        boolean decomposable = aggregationNode.isDecomposable(functionRegistry);
+
+        if (aggregationNode.getStep().equals(SINGLE) &&
+                aggregationNode.hasEmptyGroupingSet() &&
+                aggregationNode.hasNonEmptyGroupingSet()) {
+            checkState(
+                    decomposable,
+                    "Distributed aggregation with empty grouping set requires partial but functions are not decomposable");
+            return Optional.of(split(aggregationNode, context));
+        }
+
+        if (!decomposable) {
+            return Optional.empty();
+        }
+
+        // partial aggregation can only be pushed through exchange that doesn't change
+        // the cardinality of the stream (i.e., gather or repartition)
```

Comment:
This is pure migration without code changes. The purpose of this PR is to fix `PushPartialAggregationThroughJoin`